### DISCUSSION TITLE
Open a Random Video from a Specific YouTube Channel

### DISCUSSION THREAD

**Discussion by M0nica** [johnlindquist/kit#684] [Share]
I created a Script to open a random video from a specific YouTube Channel. The script prompts user for the name of the channel, uses the `scrapeSelector` to get videos from the channel and then selects a random one to open. 

https://user-images.githubusercontent.com/6998954/148695801-f5607143-265b-4fd5-aa48-de0182d36791.mov

Future enhancements: I'd like to add some error logging to help users if they enter an invalid channel name that returns a 404 or if a given channel exists but doesn't return any videos. I tried adding some console.logs but wasn't sure how to output helpful logs to ScriptKit users. 

```js
// Name: Random Video From YouTube Channel
// Description: Open a random video from a specific YouTube Channel
// Author: Monica Powell
// Twitter: @indigitalcolor
// GitHub: @m0nica

``
import "@johnlindquist/kit"


let channelName = await arg("YouTube Channel Name (e.g, yogawithadriene):")

let videos = await scrapeSelector(
    `https://www.youtube.com/c/${channelName}/videos`,
    "a#video-title",
    el => ({
        name: el.innerText,
        value: el.href,
    })
)


let randomVideo = videos[Math.floor(Math.random() * videos.length)]

exec(`open "${randomVideo.value}"`)
 ```

View gist: https://gist.github.com/M0nica/939bd4c3dfb2bf6fed08afb17ab16d19

**Comment by johnlindquist** [comment]
Sorry for taking so long to reply. This looks super fun!

Logging is typically done using `console.log()`.

If your script is in:
`~/.kenv/scripts/my-script.js`
The log will be:
`~/.kenv/logs/my-script.log`

The log should also be accessible in the options menu (the video is showing a work-in-progress built-in editor experience, for everyone else it will open in vs code or your selected editor):

https://user-images.githubusercontent.com/36073/149012918-d33ae613-af4c-47c8-9375-c8d30b48e4c1.mp4

## More Logging Options

### `inpect()`

Using `inspect(someObject)` will generate and open a text file based on the object you pass in

### `dev()`

Using `dev(someObject)` will open a Chrome dev tools where you can play "live" with the object. The object is accessible as the variable `x`, so you can try `x.length`, `x.map`, or whatever you'd normally use in dev tools

## Handling 404s in the Scraper

Here's the entire source of `scrapeSelector`. You can copy/paste the body of the function if you'd like to play around with customizing it. I'm definitely open to any suggestions. I'd have to dig into the https://playwright.dev/ docs to see what all the options are for when the selector fails to find anything.


```js
global.scrapeSelector = async (
  url: string,
  selector: string,
  xf?: (element: any) => any,
  { headless = true, timeout = 10000 } = {
    headless: true,
    timeout: 10000,
  }
) => {
  /** @type typeof import("playwright") */
  let { chromium } = await global.npm("playwright")

  if (!xf) xf = el => el.innerText
  let browser = await chromium.launch({ headless })
  let context = await browser.newContext()
  let page = await context.newPage()
  page.setDefaultTimeout(timeout)

  if (!url.startsWith("http")) url = "https://" + url
  await page.goto(url)
  // await page.waitForSelector(selector)
  let selectorHandles = await page.$$(selector)
  let results = await Promise.all(
    selectorHandles.map(async handle => {
      let result = await handle.evaluate(xf)
      await handle.dispose()
      return result
    })
  )

  await browser.close()
  return results
}
```
